woof-code/rootfs-packages/firewall_ng/usr/sbin/firewall_ng

```diff
@@ -1448,7 +1448,7 @@ else
 	exit
 fi
 
-[ "$1" = "enable" ] && exit
+["$MAIN" = "false" ] && exit
```

Comment:
Syntax error, missing space after `[`

Reply:
Yes, I now see that. However this line 1451 statement did in fact work correctly during my testing. However in the quest for proper language syntax, I will correct it.

Reply:
Interesting, when I add the extra space on the line it breaks the test function. Will have to study this further,

Reply:
Revised line 1451 with correct syntax.

Reply:
See https://github.com/puppylinux-woof-CE/woof-CE/pull/3952#issuecomment-1445280171 and https://github.com/puppylinux-woof-CE/woof-CE/pull/3952#issuecomment-1445284024

Reply:
Yeah, for some reason I did not see these when I commented afterwards.

Reply:
I tried those [https://github.com/puppylinux-woof-CE/woof-CE/pull/3952#issuecomment-1445280171 and https://github.com/puppylinux-woof-CE/woof-CE/pull/3952#issuecomment-1445284024](url) and unfortunately it still breaks firewall restart. Maybe on a reboot it works but the original functionality restarted the firewall with no further user interaction. I would think this is the preferred end user experience.

Reply:
How do you restart the firewall?

Reply:
> How do you restart the firewall?

`[ "$1" = "enable" -a "$MAIN" = "false" ] && exit` allows the code to proceed to the `urxvt -bg yellow -fg black -geometry 60X10+1+1 --hold -e /etc/init.d/rc.firewall start &` line when Apply button is pressed. If the user presses Cancel, `[ "$1" = "enable" -a "$MAIN" = "false" ] && exit`  short circuit exits and never restarts the firewall. Nor does it save any changes either. So the firewall existing state is preserved. On/Off/disabled or otherwise.

Reply:
Unfortunately, Alien Bob's  original code still creates a rc.firewall file if the user unchecks the "When this button is checked a generic firewall is generated on no specific interface." and then presses Apply. I would think most users would just press Cancel if they don't want a firewall anyway.

Reply:
If I'm not mistaken (maybe my memory from this morning is no longer correct), `MAIN=true` and not `MAIN=false` if the firewall is enabled, so clicking Apply without changing anything restarts the firewall when the firewall is already enabled. `bash -x firewall_ng` shows this.

Reply:
I temporarily inserted `gtkdialog-splash -bg yellow -timeout 2 -text "Main: $MAIN"`  one line above `[ "$1" = "enable" -a "$MAIN" = "false" ] && exit` on my local testing copy and MAIN=false when I press Apply. Have not studied the rest of the code to trace out why this would be. But `[ "$1" = "enable" -a "$MAIN" = "false" ] && exit` does work correctly for me. Using it restarts the firewall when I press Apply. And Cancel does preserve the previously "existing state". So I think this is what we are trying to accomplish here unless I am missing something.

Reply:
Did you run firewall_ng with `firewall_ng enable`? If not, this condition doesn't do anything (because it's false).

Reply:
I ran it from the menu. if that is set this way.

Reply:
I just checked my local .jwmrc file and it is just using `firewall_ng`

Reply:
which file is executing the `firewall_ng enable` command? I checked welcome1stboot.

Reply:
https://github.com/puppylinux-woof-CE/woof-CE/blob/28f7424eef841fb3591b9c989c7d5bfd25153128/woof-distro/x86_64/debian/bullseye64/_00build.conf#L132

Reply:
> I just checked my local .jwmrc file and it is just using `firewall_ng`

Did you replace /usr/sbin/firewall_ng with the modified one from this PR?

Reply:
No, but I've been editing that file for testing.

Reply:
> https://github.com/puppylinux-woof-CE/woof-CE/blob/28f7424eef841fb3591b9c989c7d5bfd25153128/woof-distro/x86_64/debian/bullseye64/_00build.conf#L132

So there is no local app that is using this command?

Reply:
No.

Reply:
I unticked this checkbox and clicked Apply:

![apply](https://user-images.githubusercontent.com/1471149/221432675-5d8c63b3-410a-433f-9605-3392f76cf792.png)

As I expect, the firewall is restarted, not disabled. I'm running `bash -x firewall_ng`:

```
+ cp -af /tmp/firewall_gen395815 /etc/init.d/rc.firewall
+ chmod 755 /etc/init.d/rc.firewall
+ rm -f /tmp/firewall_gen395815
+ '[' '' = enable -a false = false ']'
+ gtkdialog-splash -bg yellow -timeout 2 -text 'Starting firewall'
+ p=395919
```

`'[' '' = enable -a false = false ']'` is `[ "$1" = "enable" -a "$MAIN" = "false" ]`. Yes, unticking the checkbox sets `MAIN=false`. But, as I said, there's no first argument (it's `firewall_ng`, not `firewall_ng enable`), therefore `"$1" = "enable"` is false, therefore `"$1" = "enable" -a "$MAIN" = "false"` is false (in English, "the first argument is 'enable' _and_ the checkbox is unticked") . Because this condition is false, the script doesn't exit like it's supposed to, and starts the firewall again.

Reply:
So what do you advise? 

I think the more important functionality is having the Cancel and Apply buttons working correctly from the menu I.E. `firewall_ng`

This `firewall_ng enable` is not included in the ISO but `firewall_ng` is. No?

Reply:
using 'bash -x firewall_ng' and:
`if [ -f "$TMPFW" ];then

	[ $state = 1 -a "$1" != "enable" ] && /etc/init.d/rc.firewall stop && sleep 1

	echo "copying firewall"

	cp -af "$TMPFW" /etc/init.d/rc.firewall

	chmod 755 /etc/init.d/rc.firewall

	rm -f "$TMPFW"

else

	echo "Something went wrong"

	exit

fi`


using apply with:
`[ "$1" = "enable" -o "$MAIN" = "false" ] && exit`

Does not restart the firewall.

using:

`[ "$1" = "enable" -a "$MAIN" = "false" ] && exit`

Does restart the firewall.

>With the changes in this PR, the way it is now (https://github.com/puppylinux-woof-CE/woof-CE/commit/6bb4ddfe2519c31b0d0967d957b1571d2e07cf57), I unticked this checkbox and clicked Apply:

apply

As I expect, the firewall is restarted, not disabled like it's supposed to. I'm running bash -x firewall_ng:

+ cp -af /tmp/firewall_gen395815 /etc/init.d/rc.firewall
+ chmod 755 /etc/init.d/rc.firewall
+ rm -f /tmp/firewall_gen395815
+ '[' '' = enable -a false = false ']'
+ gtkdialog-splash -bg yellow -timeout 2 -text 'Starting firewall'
+ p=395919

'[' '' = enable -a false = false ']' is [ "$1" = "enable" -a "$MAIN" = "false" ]. Yes, unticking the checkbox sets MAIN=false. But, as I said, there's no first argument (it's firewall_ng, not firewall_ng enable), therefore $1 is "", an empty string. "$1" = "enable" is false, therefore "$1" = "enable" -a "$MAIN" = "false" is false (in English, "the first argument is 'enable' and the checkbox is unticked") . Because the big condition is false, the script doesn't exit like it's supposed to (the && exit part doesn't happen), and the script starts the firewall again.<

================
You will find if you run the original firewall_ng from the main repo without our mods, the untick/apply issue already exists. So it is not from this mod. It is a bug (or feature) from Alien Bob's code. I don't propose to try and resolve that. More importantly, I think fixing the apply issue is more critical at this point, No?

Reply:
> More importantly, I think fixing the apply issue is more critical at this point, No?

> Update: I think fixing the apply restarting the firewall issue is more critical at this point, No?

At this point, I'm longer sure what you mean by "the apply issue" and "restarting the firewall issue".

With the patch from https://github.com/puppylinux-woof-CE/woof-CE/pull/3952#issuecomment-1445284024, if you click Apply when the firewall is already enabled, it gets restarted, and if you click Apply after the checkbox is unticked, it's disabled. It solves the two issues I see, while this PR doesn't and even breaks `firewall_ng enable`.

Reply:
OK, I just tried this code. For some reason I did not see this until well after we started discussing the AND / OR snippet several hours later. Yes, It seems to have solved both issues. My humble apologies for not trying this earlier. I will close this PR as you have the better solution.